File: Session_05_DynamicProgramming_Basics/multi_language_implementations/coin_change.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const INF = 1e9

func getMinCoins(coins []int, amount int) int {
	dp := make([]int, amount+1)
	dp[0] = 0
	for i := 1; i <= amount; i++ {
		dp[i] = INF
	}
	for i := 1; i <= amount; i++ {
		for j := 0; j < len(coins); j++ {
			if coins[j] <= i {
				subRes := dp[i-coins[j]]
				if subRes != INF && subRes+1 < dp[i] {
					dp[i] = subRes + 1
				}
			}
		}
	}
	result := dp[amount]
	if result == INF {
		return -1
	}
	return result
}

func main() {
	coins := []int{1, 3, 4}
	amount := 6
	fmt.Printf("Min coins required: %d\n", getMinCoins(coins, amount))
	fmt.Printf("\nPress Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

File: Session_05_DynamicProgramming_Basics/multi_language_implementations/coin_change.cpp
#include <iostream>
#include <vector>

const int INF = 1e9;

int getMinCoins(const std::vector<int> &coins, int amount) {
    std::vector<int> dp(amount + 1, INF);
    dp[0] = 0;
    for (int i = 1; i <= amount; i++) {
        for (size_t j = 0; j < coins.size(); j++) {
            if (coins[j] <= i) {
                int subRes = dp[i - coins[j]];
                if (subRes != INF && subRes + 1 < dp[i]) {
                    dp[i] = subRes + 1;
                }
            }
        }
    }
    int result = dp[amount];
    return (result == INF) ? -1 : result;
}

int main() {
    std::vector<int> coins = {1, 3, 4};
    int amount = 6;
    std::cout << "Min coins required: " << getMinCoins(coins, amount) << "\n";
    std::cout << "\nPress Enter to exit...";
    std::cin.get();
    return 0;
}
